src/cryptonote_core/service_node_quorum_cop.cpp

```diff
@@ -41,165 +41,352 @@
 
 namespace service_nodes
 {
+  static_assert(quorum_cop::REORG_SAFETY_BUFFER_IN_BLOCKS < deregister_vote::VOTE_LIFETIME_BY_HEIGHT,
+                "Safety buffer should always be less than the vote lifetime");
+
   quorum_cop::quorum_cop(cryptonote::core& core)
-    : m_core(core), m_uptime_proof_height(0)
+    : m_core(core), m_uptime_proof_height(0), m_last_checkpointed_height(0)
   {
-    init();
   }
 
   void quorum_cop::init()
   {
-    m_uptime_proof_height = 0;
+    m_uptime_proof_height      = 0;
+    m_last_checkpointed_height = 0;
     m_uptime_proof_seen.clear();
+
+    uint64_t top_height;
+    crypto::hash top_hash;
+    m_core.get_blockchain_top(top_height, top_hash);
+
+    cryptonote::block blk;
+    if (m_core.get_block_by_hash(top_hash, blk))
+      process_quorums(blk);
   }
 
   void quorum_cop::blockchain_detached(uint64_t height)
   {
+    // TODO(doyle): Assumes large reorgs that are no longer possible with checkpointing
     if (m_uptime_proof_height >= height)
     {
       LOG_ERROR("The blockchain was detached to height: " << height << ", but quorum cop has already processed votes up to " << m_uptime_proof_height);
       LOG_ERROR("This implies a reorg occured that was over " << REORG_SAFETY_BUFFER_IN_BLOCKS << ". This should never happen! Please report this to the devs.");
       m_uptime_proof_height = height;
     }
+
+    if (m_last_checkpointed_height >= height)
+    {
+      LOG_ERROR("The blockchain was detached to height: " << height << ", but quorum cop has already processed votes up to " << m_last_checkpointed_height);
+      LOG_ERROR("This implies a reorg occured that was over " << REORG_SAFETY_BUFFER_IN_BLOCKS << ". This should never happen! Please report this to the devs.");
+      m_last_checkpointed_height = height;
+    }
+    m_vote_pool.remove_expired_votes(height);
   }
 
-  void quorum_cop::block_added(const cryptonote::block& block, const std::vector<cryptonote::transaction>& txs)
+  void quorum_cop::set_votes_relayed(std::vector<quorum_vote_t> const &relayed_votes)
   {
-    process_uptime_quorum(block);
-    process_checkpoint_quorum(block);
+    m_vote_pool.set_relayed(relayed_votes);
   }
 
-  void quorum_cop::process_uptime_quorum(cryptonote::block const &block)
+  std::vector<quorum_vote_t> quorum_cop::get_relayable_votes()
   {
-    uint64_t const height = cryptonote::get_block_height(block);
-    if (m_core.get_hard_fork_version(height) < 9)
-      return;
+    std::vector<quorum_vote_t> result = m_vote_pool.get_relayable_votes();
+    return result;
+  }
+
+  static int find_index_in_quorum_group(std::vector<crypto::public_key> const &group, crypto::public_key const &my_pubkey)
+  {
+    int result = -1;
+    auto it = std::find(group.begin(), group.end(), my_pubkey);
+    if (it == group.end())
+      return result;
+    result = static_cast<int>(it - group.begin());
```

Comment:
could rework as (and avoid the `static_cast`):

```c++
auto it = std::find(group.begin(), group.end(), my_pubkey);
if (it == group.end()) { return -1; }
return std::distance(group.begin(), it);
```

Reply:
Added